### PR TITLE
[feat][broker] PIP-157 Bucketing topic metadata to allow more topics per namespace

### LINKED ISSUE #15254: PIP-157: Bucketing topic metadata to allow more topics per namespace
## Motivation

Pulsar is able to manage millions of topics but the number of topics within a single namespace is limited by metadata storage.

For each topic within a namespace there is a ZooKeeper node. Listing topics thus requires listing children of a node, which at around 10K topics hits the limits of ZK.

## Goal

This feature will allow larger number of topics within a namespace by inserting an intermediate layer (buckets) before the topic nodes like `/managed-ledgers/tenant/namespace/domain/bucket/topic`.

By default this feature will be switched off and would only be enabled on a per namespace basis at the creation of namespaces by setting a policy. This eliminates the need for migrating existing installations to this new scheme.

Buckets will not have correlation with bundles.

## API Changes

A new policy `numberOfTopicBuckets` will be added. The default value, 1 means no bucketing, the current behaviour will be preserved for the namespace. Greater values mean topics will be stored at a path including buckets. Users will not be able to change the number of buckets after the namespace is created.

## Implementation 

The goal is to implement this feature transparently to the user. Clients will continue to refer to topics by domain://tenant/namespace/topic but pulsar will internally translate to the new persistence naming where necessary.

The way metadata stores work will not be affected either.

Assigning topics to buckets will be based on the topic name&#39;s hash code&#39;s absolute value modulo the number of buckets.

The bulk of the changes necessary for this feature is to make namespace policies available wherever persistence naming is calculated. Where listing of topics within a namespace is necessary, the introduction of the new layer will add some overhead in the form of multiple requests to the metadata store. These include checking checking if the limit on topic number per namespace has been reached.

### Example

Let&#39;s consider the following metadata hierarchy:

```
managed-ledgers
\-  tenant
    \-  namespace
        \-  persistent
            +-  nptopic1
            +-  nptopic2
            +-  ptopic-partition-0
            +-  ptopic-partition-1
            +-  ptopic-partition-2
            \-  ptopic-partition-3
```

In case of 3 buckets the same topic metadata would be laid out the following way:

```
managed-ledgers
\-  tenant
    \-  namespace
        \-  persistent
            +-  $0
            |   +-  ptopic-partition-0
            |   \-  ptopic-partition-3
            +-  $1
            |   +-  nptopic2
            |   \-  ptopic-partition-1
            \-  $2
                +-  nptopic1
                \-  ptopic-partition-2
```

### Compatibility

Existing namespaces and namespaces created without explicitly activating this feature will not be affected.

Namespaces created with this feature activated can be used just as others.

## Rejected alternatives

An alternative approach would be to introduce bucketing globally for all namespaces. This would make metadata structure more homogeneous but would require complex update logic to atomically move topics from their current path to the new place once all brokers are upgraded.
For similar reasons changing the number of buckets is not a goal of this proposal.

Since the proposal intends to solve a problem related to ZK, it could be handled within the ZK based metadata store implementation. This would have to introduce knowledge of  what paths mean thus breaking separation of concerns.

### PR DESCRIPTION
**Pull request by andrasbeni**
Fixes #15254

### Motivation

The number of topics in a namespace is limited by ZooKeeper when it is used as the local store. This is because listing topics requires all topic names to fit in ZK's buffer. By introducing an intermediate layer between the namespace node and topics, a namespace can have more topics.

### Modifications

 - Add new namespace policy `number_of_buckets` defaulting to 1 (i.e. no buckets)
 - Calculate topics' persistence name based on the number of buckets in the namespace

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

*(example:)*
  - Added integration tests for creating topics in a bucketed namespace and producing to / consuming from it

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [X] `doc-required` 
(Your PR needs to update docs and you will update later)

### Matching PR in forked repository

PR in forked repository: https://github.com/andrasbeni/pulsar/pull/4

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java [hunk 1]
```diff
@@ -1219,7 +1220,11 @@ private CompletableFuture<List<String>> getPartitionsForTopic(TopicName topicNam
     }
 
     public CompletableFuture<List<String>> getListOfPersistentTopics(NamespaceName namespaceName) {
-        return pulsar.getPulsarResources().getTopicResources().listPersistentTopicsAsync(namespaceName);
+        return pulsar.getPulsarResources().getTopicResources().listPersistentTopicsAsync(namespaceName)
+                .thenApply(x -> {
+                    LOG.info("getListOfPersistentTopics: " + x);
```
[the remark this answers is not in the document]
Comment (andrasbeni) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java]: I don't think it needs to be logged at all. Reverted this part.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/TopicResources.java [hunk 2]
```diff
@@ -41,22 +42,29 @@ public class TopicResources {
...
-        return store.getChildren(path).thenApply(children ->
-                children.stream().map(c -> TopicName.get(TopicDomain.persistent.toString(), ns, decode(c)).toString())
-                        .collect(Collectors.toList())
-        );
+        CompletableFuture<List<String>> childrenOfNamespace = store.getChildren(path);
+        CompletableFuture<List<String>> topics = childrenOfNamespace.thenCompose(children ->
+                !children.isEmpty() && children.get(0).startsWith(TopicName.BUCKET_ID_PERFIX)
+                        ? children.stream().map(bucket -> store.getChildren(path + "/" + bucket))
+                            .collect(FutureUtil.toList())
+                        : CompletableFuture.completedFuture(children));
```
[the remark this answers is not in the document]
Comment (andrasbeni) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/TopicResources.java:71]: Nice catch. I did not realize this. I extracted the common part into a method.

**Comment by andrasbeni** [comment]
@merlimat , please take a look

**Comment by andrasbeni** [comment]
@merlimat, I don't think I will have time to work on this anytime soon. Feel free to take over if this is still relevant.

**Comment by tisonkun** [comment]
Thanks for your contribution @andrasbeni!

It seems the community doesn't give this PR a review in time. I add this PR in my backlog to see if it's still relevant. Perhaps ping you for reviewing if so.

**Pull request closed without merging**